pyffs/czt.py

```diff
@@ -121,7 +122,139 @@ def czt(x, A, W, M, axis=-1):
     G *= V
     g = fftpack.ifft(G, axis=axis)
     g_mod = np.float_power(W, (n[:M] ** 2) / 2)
-    g[_util._index(g, axis, slice(M))] *= g_mod.reshape(sh_M)
+    g[_index(g, axis, slice(M))] *= g_mod.reshape(sh_M)
 
-    X = g[_util._index(g, axis, slice(M))]
+    X = g[_index(g, axis, slice(M))]
     return X
+
+
+def czt2(Phi, Ax, Ay, Wx, Wy, Mx, My, axes=(-2, -1)):
+    return cztn(Phi, A=[Ax, Ay], W=[Wx, Wy], M=[Mx, My], axes=axes)
+
+
+def cztn(Phi, A, W, M, axes=None):
+    """
+    Perform multi-dimensional CZT from signal samples, using the multi-dimensional FFT.
+
+    Parameters
+    ----------
+    Phi : :py:class:`~numpy.ndarray`
+        (..., N_s1, N_s2, ..., N_sD, ...) input values.
+    A : list(float or complex)
+        Circular offset from the positive real-axis, for each dimension.
+    W : list(float or complex)
+        Circular spacing between transform points, for each dimension.
+    M : list(int)
+        Length of transform for each dimension.
+    axes : tuple
+        Dimensions of `Phi` along which transform should be applied.
+
+    Returns
+    -------
+    X : :py:class:`~numpy.ndarray`
+        (..., M_1, M_2, ..., M_D, ...) transformed input along the axis indicated by `axes`.
+
+    Notes
+    -----
+    Due to numerical instability when using large `M`, this implementation only supports transforms
+    where `A` and `W` have unit norm.
+    """
+
+    axes, A, W = _verify_cztn_input(Phi, A, W, M, axes)
+
+    # Initialize variables
+    D = len(axes)
+    N = np.array(Phi.shape)[axes]
+    L = []
+    n = []
+    for d in range(D):
+        _L = fftpack.next_fast_len(N[d] + M[d] - 1)
+        L.append(_L)
+        n.append(np.arange(_L))
+
+    # Initialize input
+    sh_U = list(Phi.shape)
+    for d in range(D):
+        sh_U[axes[d]] = L[d]
+    u_dtype = (
+        np.complex64
+        if ((Phi.dtype == np.dtype("complex64")) or (Phi.dtype == np.dtype("float32")))
+        else np.complex128
+    )
+    u = np.zeros(sh_U, dtype=u_dtype)
+    idx = _index_n(u, axes, [slice(n) for n in N])
+    u[idx] = Phi
+
+    # Modulate along each dimension
+    for d in range(D):
+        _n = n[d]
+        _N = N[d]
+        sh_N = [1] * Phi.ndim
+        sh_N[axes[d]] = N[d]
+        u_mod_d = (A[d] ** -_n[:_N]) * np.float_power(W[d], (_n[:_N] ** 2) / 2)
+        u[idx] *= u_mod_d.reshape(sh_N)
```

Comment:
Why `u[idx]` instead of plain `u`?

Reply:
It's the same idea as [here](https://github.com/imagingofthings/pyFFS/blob/769d6a9bebeffa228730d0a7a516b9371e65ea7c/pyffs/__init__.py#L368) in the 1D case. I think it's because `u` (or `y` in the 1D case) is zero-padded. So there's no need to modulate the zeros samples.

Reply:
Also raises an error because broadcasting cannot be performed.